driver-core/src/main/com/mongodb/internal/client/model/AbstractConstructibleBson.java

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.mongodb.internal.client.model;
+
+import com.mongodb.annotations.Immutable;
+import org.bson.BsonDocument;
+import org.bson.BsonValue;
+import org.bson.codecs.configuration.CodecRegistry;
+import org.bson.conversions.Bson;
+
+import java.util.Objects;
+
+/**
+ * A {@link Bson} that allows constructing new instances via {@link #newAppended(String, BsonValue)} instead of mutating {@code this}.
+ * While instances are not {@link Immutable immutable},
+ * {@link BsonDocument#isEmpty() empty} instances are treated specially and are immutable.
+ *
+ * @param <S> A type introduced by the concrete class that extends this abstract class.
+ */
+public abstract class AbstractConstructibleBson<S extends AbstractConstructibleBson<S>> implements Bson {
+    private final BsonDocument doc;
+
+    protected AbstractConstructibleBson(final Bson doc) {
+        this.doc = doc.toBsonDocument();
```

Comment:
This issue is related to https://github.com/mongodb/mongo-java-driver/pull/891#discussion_r842255622. I agree, that's the right thing to do, and it appears to be doable. I will refactor `AbstractConstructibleBson`, `AbstractConstructibleBsonElement`, `SearchFacet.numberFacet`/`dateFacet`, `SearchPath.toBsonValue` and the places where it is used.

Reply:
Changing all new types to be rendered into `BsonDocument` lazily, which allows us to wait and use `documentClass` and `codecRegistry` once they are available, is not exactly straightforward, and I have not completed it yet.

Also, this cannot be done for `SearchPath` at all because we can only render lazily `Bson`, and turns out, that using the form `"fieldPath"` (BSON String) instead of `{"value": "fieldPath"}` is mandatory, not optional. This means that `SearchPath` is not `Bson` (some of its forms are documents, but some are just strings), i.e., cannot be rendered lazily.

Reply:
Done, everything is now encoded to `BsonDocument` lazily except for `SearchPath`, because it can't always be represented as a document (sometimes it must be represented as a string), and the driver's codec system supports lazy encoding only for documents via `Bson`.